### DISCUSSION TITLE
DataGridRow.Cells need for it to be internal?

### DISCUSSION THREAD

**Discussion by alaingaudreau** [AvaloniaUI/Avalonia#8596] [Q&A]
Hi, 

I'm trying to understand the need for DataGridRow.Cells to be internal.

I haven't studied the Avalonia code much but from an answer I got from @maxkatz6  yesterday regarding styling a row based on bound item's property, I was looking at what we can do from within with a DataGrid.LoadingRow event and can't see a simple way to reach a DataGridCell within that row.

When LoadingRow is fired, the visual tree doesn't seem to have been populated yet so we can't access DataGridCell through e.Row?.FindDescendantOfType < DataGridCellsPresenter > ().etc...

The idea behind it would be to be able to change a DataGridCell's class or padding property in order for example to add left padding on it's TextBlock to offset it's text from the row above's TextBlock.

So turning the first column into a basic TreeView visual layout based on that same property's value and not affecting the rest of the columns in the grid.

The sender in the event being a DataGrid, I figured we may have a way to go through it's model in some manner like datagrid.Rows[DataGridRowEventArgs.Rows.Index].Columns[] or .Cells[]  but that's not possible either.

**Comment by maxkatz6** [comment]
Cells, the same as Rows, can be reused. I.e. single cell might pop up in different places of the grid while scrolling. I suppose it was the reason why Silverlight developers made it internal (we ported this DataGrid from there). As well as on horizontal scrolling, when rows are not changed. 

As an alternative, you can set Row.Classes inside of the LoadingRow, and combine it with DataGridTextColumn.StyleClasses. It would allow to use style selectors like "DataGridRow.special-row DataGridCell.special-column-cell". 

Or, you can use DataGridTemplatedCell, where you have more control over the cell content.

**Reply by alaingaudreau** [reply]
From looking at DataGrid.cs and DataGridRow.cs, Cells and FillerCell property are barely being used and FillerCell has a //TODO Styles on the internal FillerCell

Wonder if someone had the idea of allowing a certain control on the cell styling at some point through either of those classes?

I tried using DataGridRow.Classes.Add() to set one of 6 classes from within LoadingRow but it's really not working well.

The datagrid rendering is rather inconsistent using classes/styles. If you scroll down the grid with the scroll bar, the style isn't getting applied every time. If you scroll up and down the styles will move around, some rows getting styled some not then the next time around other rows are affected.

Screenshot-1 is done using reflection to set DataGridCell.Padding and DataGridCell.FontWeight from within the LoadingRow event handler and gives the proper result every time

Screenshot-2 is done using Row.Classes.Add("someclass"); and a <Style> block per class.

Same code base with the exception of the following setCellInfo() being my method that uses reflection to set the DataGridRow.Cells._cells with a modified array of DataGridCell :

```
setCellInfo(e.Row, FontWeight.Bold, new Thickness(20, 0, 0, 0));
//e.Row.Classes.Add("subgroups");
```

![Screenshot-1](https://user-images.githubusercontent.com/46643341/181663555-7d152078-0483-40b7-b0fe-7630cb25fca3.png)
![Screenshot-2](https://user-images.githubusercontent.com/46643341/181663557-512182ec-c09f-4066-a00b-ef8c80fb3e9d.png)

**Reply by timunie** [reply]
Would it probably be an option to set a cell style in the column itself for each cell, but to set Padding or other Properties via Binding and IValueConverter? 

At least this works for me (Padding only): 
![image](https://user-images.githubusercontent.com/47110241/181695631-1ff8e4dd-718e-4b3b-8ddd-9632a094b0a8.png)

*Note: I had to hide most content as the project plan is confidental*

### How does it work? 

As my Cell is much more complex, I have my own ControlTemplate. But should also work with simple Setters similar, 

```xml
    <Styles.Resources>
        <x:Double x:Key="TreeViewItemIndent">16</x:Double>
        <converters:MarginMultiplierConverter Indent="{StaticResource TreeViewItemIndent}" Left="True"
                                              x:Key="TreeViewItemLeftMarginConverter" />
    </Styles.Resources>

    <Style Selector="DataGridCell.treeRoot">
        <Setter Property="Template">
            <ControlTemplate>
             [...]
            <!-- Level is a property in my Item-VM -->
            <Grid ColumnDefinitions="Auto, *" x:Name="PART_ContentRoot" Grid.Column="0"
                          Margin="{Binding Level, Mode=OneWay, Converter={StaticResource TreeViewItemLeftMarginConverter}}">
             [...]
```
 
the Converter: 

```cs
    public class MarginMultiplierConverter : IValueConverter
    {
        public double Indent { get; set; }

        public bool Left { get; set; } = false;

        public bool Top { get; set; } = false;

        public bool Right { get; set; } = false;

        public bool Bottom { get; set; } = false;

        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            if (value is int scalarDepth)
            {
                return new Thickness(
                    Left ? Indent * scalarDepth : 0,
                    Top ? Indent * scalarDepth : 0,
                    Right ? Indent * scalarDepth : 0,
                    Bottom ? Indent * scalarDepth : 0);
            }
            else if (value is Thickness thicknessDepth)
            {
                return new Thickness(
                    Left ? Indent * thicknessDepth.Left : 0,
                    Top ? Indent * thicknessDepth.Top : 0,
                    Right ? Indent * thicknessDepth.Right : 0,
                    Bottom ? Indent * thicknessDepth.Bottom : 0);
            }
            return new Thickness(0);
            
        }

        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new System.NotImplementedException();
        }
    }
```
       

Happy coding
Tim

**Reply by alaingaudreau** [reply]
Thanks @timunie for the suggestion with your working implementation.

That seems to work quite well.

I just set up a quick example to try it out and it seems to work reliably, much more so than setting the classes through DataGrid.LoadingRow was giving me.

It adds greatly to the complexity of the code but at least it works and doesn't require breaking encapsulation.

For anyone following this thread, it works with simple Style and Setters without the template as well.

**Reply by timunie** [reply]
Let me add one more hint for the next time: if you have a specific problem to solve, try to make clear for you and for others what's the final result should be. It's sometimes not an easy task and it is also not to criticize you, it just can help to see other solutions than the one you originally tried. 

Have a great weekend 😊
